Document Tier 1 support for OpenShift Authoring HA template (#81)
mramendi/kie-docs: doc-content/enterprise-only/openshift/environment-authoring-con.adoc

[id='environment-authoring-con'] 
ifdef::PAM[]
= Authoring environment
You can deploy an environment for creating and modifying processes using {CENTRAL}. It consists of {CENTRAL} for the authoring work and {KIE_SERVER} for test execution of the processes.
endif::PAM[]
ifdef::DM[]
= Authoring or managed server environment
You can deploy an environment for creating and modifying services using {CENTRAL} and for running them in {KIE_SERVERS} managed by {CENTRAL}. This environment consists of {CENTRAL} and one or more {KIE_SERVERS}.

> You can use {CENTRAL} both to develop services and to deploy them to one or several {KIE_SERVERS}. For example, you can deploy test versions of services to one {KIE_SERVER} and production versions to another {KIE_SERVER}.

Reviewer: I think that this line is not needed. I am actually more confused. If am I understand this line, it basically says that we can connect more KIE Servers to one Central, right?

Author: Yes, but we gotta offer a use case for that...

> To avoid accidentally deploying wrong versions to a production {KIE_SERVER}, you can create separate environments to author services (_authoring environment_) and to manage deployment of production services (_managed server environment_). You can use a shared external Maven repository between these environments, so that services developed in the authoring environment are available in the managed server environment. Usually, one {KIE_SERVER} is sufficient for an authoring environment.

Reviewer: The description is too complicated, it would be good to have there some image where is illustrated this env.
Also to avoid deploying wrong version there is parameter KIE_SERVER_MODE for KIE Server, on manged template this is by default set to "PRODUCTION" so it will not allow to deploy snapshot kjars. [See DM KIE Server template](https://github.com/jboss-container-images/rhdm-7-openshift-image/blob/master/templates/rhdm77-kieserver.yaml#L119) in RHDM auth template this mode is set to "DEVELOPMENT"

Author: We have an additional task to do a proper architectural overview, Adriel is currently working on the input. We will try to get a diagram in as a part of that; we don't have the complete information for a diagram right now.

Author: (To be clear, that task is probably getting done for 7.7 not 7.6)

> For {PRODUCT}, the procedures to deploy an authoring environment and a managed server environment are the same. You must first deploy an authoring environment template, consisting of {CENTRAL} and one {KIE_SERVER}. Then, if necessary, you can deploy additional {KIE_SERVER} templates in the same namespace to create a managed server environment with multiple {KIE_SERVERS}.

Reviewer: I would say that we do not needed to talk about managed environment for RHDM at all. Managed environment is in RHPAM where BC Monitoring can monitor running process and task. In case of RHDM we do not have any process or tasks. I have the feeling that we are more likely talking here about development vs. production environment.

Author: "Managed" is simply the term for using BC to deploy services across KIE servers. I don't want to call it production as I don't want to tell the user how many levels of staging they should have.

endif::DM[]

Depending on your needs, you can deploy either a single authoring environment template or a high-availability (HA) authoring environment template.

A single authoring environment contains two pods. One of the pods runs {CENTRAL}, the other runs {KIE_SERVER}. 

> ifdef::PAM[The {KIE_SERVER} includes an embedded H2 database engine.]

Reviewer: Or any other database engine, users can replace dbs

Author: Will change to "by default". They can replace DBs but they have to modify the template to do that.

> This type of environment uses the least possible amount of resources. 

Reviewer: This is not valid for Authoring. This sentence is true for Trial templates. Also users can set BC and KS limits by env. variables e.g. by DECISION_CENTRAL_MEMORY_LIMIT

Author: I will modify to make it clear this is the least possible amount *for full-fledged authoring*. In Trial you can't save your work.

> An HA authoring environment contains several pods. Both {CENTRAL} and {KIE_SERVER} are provided in scalable pods that can run in parallel and share persistent storage. 

Reviewer: ```suggestion
In HA authoring environment both {CENTRAL} and {KIE_SERVER} are provided in scalable pods that run in parallel and have share persistent storage. 
```

ifdef::PAM[The database is provided by a separate pod.]

> Additional pods also contain the AMQ and Data Grid servers that enable high-availability functionality in {CENTRAL}. Use a high-availability authoring environment to provide maximum reliability and responsiveness, especially if several users are involved in authoring at the same time.

Reviewer: ```suggestion
To enable high-availability functionality in {CENTRAL} additional pods with AMQ and Data Grid are required. These pods are configured and deployed by templates or operator. Use a high-availability authoring environment to provide maximum reliability and responsiveness, especially if several users are involved in authoring at the same time.
```
Not sure if this sounds better, but it was not clean for me from where these AMQ and Data Grid pods are.

In the current version of {PRODUCT}, an HA authoring environment is supported with certain limitations:

> * If one user is deploying a service to the {KIE_SERVER}, another user will not be able to deploy a service to the same {KIE_SERVER} until the first deployment completes.

Reviewer: This limitation is not only for HA but for whole OpenShiftStartupStrategy

Author: HA is when the user is likely to feel it.

* If a {CENTRAL} pod crashes while a user works with it, the user can get an error message and then is redirected to another pod. Logging on again is not required. 

> * If a {CENTRAL} pod crashes during a user operation, data that was not committed might be lost. 

Reviewer: ```suggestion
* If a {CENTRAL} pod crashes during a user operation, data that was not committed (saved) might be lost. 
```

* If a {CENTRAL} pod crashes during creation of a project, an unusable project might be created. 

* If a {CENTRAL} pod crashes during creation of an asset, the asset might be created but not indexed, so it can not be used. The user can open the asset in {CENTRAL} and save it again to make it indexed.

ifdef::PAM[]

> You can also deploy additional managed or immutable {KIE_SERVERS}, if required. {CENTRAL} can automatically discover any {KIE_SERVERS} in the same namespace, including immutable {KIE_SERVERS} and managed {KIE_SERVERS}. This feature requires the `OpenShiftStartupStrategy` setting, which is enabled for all {KIE_SERVERS} except those deployed in a fixed managed infrastructure. For instructions about deploying managed {KIE_SERVERS} with the `OpenShiftStartupStrategy` setting enabled, see {URL_DEPLOYING_MANAGED_FREEFORM_ON_OPENSHIFT}[_{DEPLOYING_MANAGED_FREEFORM_ON_OPENSHIFT}_]. For instructions about deploying immutable {KIE_SERVERS}, see {URL_DEPLOYING_IMMUTABLE_ON_OPENSHIFT}[_{DEPLOYING_IMMUTABLE_ON_OPENSHIFT}_].

Reviewer: This is also possible for DM

Author: The DM version mentions the possibility elsewhere in the text and, more importantly, this entire part about document links does not apply. I'll see if I can do some more PAM/DM unification here; before this version I had separate files in place for PAM and DM for this section.

endif::PAM[]
ifdef::DM[]
= Authoring or managed server environment
You can deploy an environment for creating and modifying services using {CENTRAL} and for running them in {KIE_SERVERS} managed by {CENTRAL}. This environment consists of {CENTRAL} and one or more {KIE_SERVERS}.

You can use {CENTRAL} both to develop services and to deploy them to {KIE_SERVERS}. You can connect several {KIE_SERVERS} to one {CENTRAL} to manage deployment of services to each of the servers. 

If necessary, you can create separate environments, so that you can use one deployment of {CENTRAL} to author services (_authoring environment_) and another deployment of {CENTRAL} to manage deployment of staging or production services on several {KIE_SERVERS} (_managed server environment_). Usually, one {KIE_SERVER} is sufficient for a dedicated authoring environment. You can use an external Maven repository to store services from an authoring environment and deploy them to a separate managed server environment.

> For {PRODUCT}, the procedures to deploy an authoring environment and a managed server environment are the same. You must first deploy an authoring environment template, consisting of {CENTRAL} and one {KIE_SERVER}. Then, if necessary, you can deploy additional {KIE_SERVER} templates in the same namespace to create an environment  with multiple {KIE_SERVERS}.

Reviewer: ```suggestion
For {PRODUCT}, the procedures to deploy an authoring environment and a managed server environment are the same. You must first deploy an authoring environment template, consisting of {CENTRAL} and one {KIE_SERVER}. Then, if necessary, you can deploy additional {KIE_SERVER} templates in the same namespace to create an environment with multiple {KIE_SERVERS}.
```
removed one space

> For {PRODUCT}, the procedures to deploy an authoring environment and a managed server environment are the same. You must first deploy an authoring environment template, consisting of {CENTRAL} and one {KIE_SERVER}. Then, if necessary, you can deploy additional {KIE_SERVER} templates in the same namespace to create an environment  with multiple {KIE_SERVERS}.

Reviewer: I think we should also mentioned there what is different between DM authoring env. and managed env.

Author: I deleted this difference on your review, I'll re-add.

endif::DM[]

Depending on your needs, you can deploy either a single authoring environment template or a high-availability (HA) authoring environment template.

A single authoring environment contains two pods. One of the pods runs {CENTRAL}, the other runs {KIE_SERVER}. 
ifdef::PAM[The {KIE_SERVER} by default includes an embedded H2 database engine.]
This environment is most suitable for single-user authoring or when your OpenShift infrastructure has limited resources. 

In an HA authoring environment, both {CENTRAL} and {KIE_SERVER} are provided in scalable pods. When pods are scaled, persistent storage is shared between the copies. 
ifdef::PAM[The database is provided by a separate pod.]
To enable high-availability functionality in {CENTRAL}, additional pods with AMQ and Data Grid are required. These pods are configured and deployed by the high-availability authoring template. Use a high-availability authoring environment to provide maximum reliability and responsiveness, especially if several users are involved in authoring at the same time.

In the current version of {PRODUCT}, an HA authoring environment is supported with certain limitations:

> * If one user is deploying a service to the {KIE_SERVER}, another user will not be able to deploy a service to the same {KIE_SERVER} until the first deployment completes. 

Reviewer: Could be add there steps or link to somewhere (maybe RH solution) how to change it to the old startup strategy so user can make changes if they needed.

Author: Documenting a template change with adding about 15 lines, as this case it, to production doc standards is not an easy task. Especially because the templates themselves are already finalized. 

What I would want to do is just put in the additions into the template but comment them out. and tell the user in the doc to uncomment the changes. Or at least place a comment where the lines need to be added, and tell the user to find the comment and add the lines.

None of this seems viable, so how would you want me to do it?

Reviewer: Maybe the best for now will be to leave the doc as it is and if someone ask just refer them to RHDM-1151. After we have fix for this issue we can provide documentation how to set it properly.

Author: Good, leaving as is, just modifying the limitation to show it does not just apply to HA.

= Authoring or managed server environment
You can deploy an environment for creating and modifying services using {CENTRAL} and for running them in {KIE_SERVERS} managed by {CENTRAL}. This environment consists of {CENTRAL} and one or more {KIE_SERVERS}.

You can use {CENTRAL} both to develop services and to deploy them to {KIE_SERVERS}. You can connect several {KIE_SERVERS} to one {CENTRAL} to manage deployment of services to each of the servers. 

If necessary, you can create separate environments, so that you can use one deployment of {CENTRAL} to author services (_authoring environment_) and another deployment of {CENTRAL} to manage deployment of staging or production services on several {KIE_SERVERS} (_managed server environment_). Usually, one {KIE_SERVER} is sufficient for a dedicated authoring environment. You can use an external Maven repository to store services from an authoring environment and deploy them to a separate managed server environment.

For {PRODUCT}, the procedures to deploy an authoring environment and a managed server environment are the same. You must first deploy an authoring environment template, consisting of {CENTRAL} and one {KIE_SERVER}. 

> Then, if necessary, you can deploy additional {KIE_SERVER} templates in the same namespace to create an environment with multiple {KIE_SERVERS}. This environment can be a managed server environment for staging and production deployment of services. 

Reviewer 2: Just noting the "Then, ..." that IBM SG says to avoid. Can just exclude if you want, "If necessary, you can deploy ..." I don't care personally, just reminding of IBM SG aversion to starting sentences with "Then" and with the form, "do x, then do y." They say that it should always be "and then". Anyway, trivial.

This environment is most suitable for single-user authoring or when your OpenShift infrastructure has limited resources. 

In an HA authoring environment, both {CENTRAL} and {KIE_SERVER} are provided in scalable pods. When pods are scaled, persistent storage is shared between the copies. 
ifdef::PAM[The database is provided by a separate pod.]
To enable high-availability functionality in {CENTRAL}, additional pods with AMQ and Data Grid are required. These pods are configured and deployed by the high-availability authoring template. Use a high-availability authoring environment to provide maximum reliability and responsiveness, especially if several users are involved in authoring at the same time.

In the current version of {PRODUCT}, an HA authoring environment is supported with certain limitations:

* If a {CENTRAL} pod crashes while a user works with it, the user can get an error message and then is redirected to another pod. Logging on again is not required. 

* If a {CENTRAL} pod crashes during a user operation, data that was not committed (saved) might be lost. 

* If a {CENTRAL} pod crashes during creation of a project, an unusable project might be created. 

> * If a {CENTRAL} pod crashes during creation of an asset, the asset might be created but not indexed, so it can not be used. The user can open the asset in {CENTRAL} and save it again to make it indexed.

Reviewer 2: "cannot" one word. Search and replace all if applicable.
